Better error messages when reflinking directories

konstin opened the pull request:
On macos, `clonefile` can handle directories (https://www.manpagez.com/man/2/clonefile/, https://opensource.apple.com/source/xnu/xnu-3789.21.4/bsd/man/man2/clonefile.2.auto.html - sorry, i couldn't find any better source online), which is a lot faster than manually walking the directory tree. This currently leads to a confusing error messages when the target directory exists, saying just `the source path is not an existing regular file` and not reporting the real cause, `File exists (os error 17)`.

This PR makes two changes: We return the original error for both files and directories, and for the error for irregular src paths, we still display the original error so the information isn't lost.

NobodyXu commented on src/lib.rs:
```diff
@@ -44,6 +44,7 @@ use std::path::Path;
 ///
```
Can you change line 43 to "MacOS / OS X / iOS" please?

NobodyXu commented on src/lib.rs:
```diff
@@ -101,12 +102,12 @@ pub fn reflink_or_copy(from: impl AsRef<Path>, to: impl AsRef<Path>) -> io::Resu
 
 fn check_is_file_and_error(from: &Path, err: io::Error) -> io::Error {
-    if from.is_file() {
+    if fs::symlink_metadata(from).map_or(false, |m| m.is_file() || m.is_dir()) {
```
Since only macos support reflink dir, we should only check for `is_dir` on MacOS/iOS

```suggestion
    if fs::symlink_metadata(from).map_or(false, |m| m.is_file() || (cfg!(any(target_os = "macos", target_os = "ios")) && m.is_dir())) {
```

NobodyXu replied:
Also it seems like that on MacOS clonefile also supports symlink according to https://www.manpagez.com/man/2/clonefile/

NobodyXu commented on tests/reflink.rs:
```diff
@@ -57,11 +76,31 @@ fn reflink_src_is_dir() {
     }
 }
 
+#[test]
```
should be macos/ios only

konstin replied:
I added a check that we show the correct error on both macos (target exists) and not macos (invalid input)

konstin commented:
Windows, macos and linux are now all passing on CI. The test should have good coverage over the matrix of { input is file, input is dir, input is symlink } x { target exists, target doesn't exist } x { windows, linux (and implicitly, other non-macos unix), macos).

NobodyXu commented on src/lib.rs:
```diff
@@ -101,12 +102,17 @@ pub fn reflink_or_copy(from: impl AsRef<Path>, to: impl AsRef<Path>) -> io::Resu
 }
```
I think we need to improve doc for `reflink_or_copy` and its implementation.

It fallbacks to [`fs::copy`](https://doc.rust-lang.org/std/fs/fn.copy.html) which can handle file but not symlink or dir.

However it calls `check_is_file_and_error`, so I think it makes sense to move the new error checking logic into `reflink` and keep the old one in `reflink_copy`

konstin commented:
I've split the logic for `reflink` and `reflink_or_copy` and added platform specific docs. I dislike the asymmetry in `reflink_or_copy`, i think we'd eventually want a `reflink_or_copy_dir_all` function, but that's out of scope of this PR.

NobodyXu commented on tests/reflink.rs:
```diff
@@ -41,13 +41,36 @@ fn reflink_dest_is_dir() {
+    let dest_file_path = dir.path().join("dest.txt");
+
+    match reflink(symlink, dest_file_path) {
+        Ok(()) => panic!(),
```
Could use `.unwrap_err()` here

NobodyXu commented on tests/reflink.rs:
```diff
@@ -57,7 +80,29 @@ fn reflink_src_is_dir() {
         Ok(()) => panic!(),
```
Same here, `.unwrap_err()`

NobodyXu commented on tests/reflink.rs:
```diff
@@ -57,7 +80,29 @@ fn reflink_src_is_dir() {
+    let src_file_path = dir.path().join("src");
+    let dest_file_path = dir.path().join("dest");
+
+    let _src_file = fs::create_dir(&src_file_path).unwrap();
+    let _dest_file = fs::create_dir(&dest_file_path).unwrap();
```
```suggestion
    fs::create_dir(&src_file_path).unwrap();
    fs::create_dir(&dest_file_path).unwrap();
```

NobodyXu commented on tests/reflink.rs:
```diff
@@ -57,7 +80,29 @@ fn reflink_src_is_dir() {
+    let _dest_file = fs::create_dir(&dest_file_path).unwrap();
+
+    match reflink(&src_file_path, &dest_file_path) {
+        Ok(()) => panic!(),
```
Could use `.unwrap_err()` here

NobodyXu commented:
> i think we'd eventually want a `reflink_or_copy_dir_all` function, but that's out of scope of this PR.

I'd welcome another PR for that and then I can cut a new release for reflink-copy.

NobodyXu commented:
Thank you!
